broker.go

```diff
@@ -474,51 +468,80 @@ func (b *Broker) offset(topic string, partition int32, timems int64) (offset int
 				},
 			},
 		},
-	})
+	}
 
-	if err != nil {
-		if _, ok := err.(*net.OpError); ok || err == io.EOF || err == syscall.EPIPE {
-			// Connection is broken, so should be closed, but the error is
-			// still valid and should be returned so that retry mechanism have
-			// chance to react.
-			log.Debug("connection died while sending message",
-				"topic", topic,
-				"partition", partition,
-				"error", err)
-			conn.Close()
+	var resErr error
+	retry := &backoff.Backoff{Min: b.conf.LeaderRetryWait, Jitter: true}
+offsetRetryLoop:
+	for try := 0; try < b.conf.LeaderRetryLimit; try++ {
+		if try != 0 {
+			time.Sleep(retry.Duration())
 		}
-		return 0, err
-	}
-	found := false
-	for _, t := range resp.Topics {
-		if t.Name != topic {
-			log.Debug("unexpected topic information",
-				"expected", topic,
-				"got", t.Name)
-			continue
+
+		conn, err := b.leaderConnection(topic, partition)
+		if err != nil {
+			return 0, err
 		}
-		for _, part := range t.Partitions {
-			if part.ID != partition {
-				log.Debug("unexpected partition information",
-					"topic", t.Name,
-					"expected", partition,
-					"got", part.ID)
+		defer func() { go b.conns.Idle(conn) }()
+
+		resp, err := conn.Offset(req)
+		if err != nil {
+			if _, ok := err.(*net.OpError); ok || err == io.EOF || err == syscall.EPIPE {
+				log.Debug("connection died while sending message",
+					"topic", topic,
+					"partition", partition,
+					"error", err)
+				conn.Close()
+				resErr = err
 				continue
 			}
-			found = true
-			// happens when there are no messages
-			if len(part.Offsets) == 0 {
-				offset = 0
-			} else {
-				offset = part.Offsets[0]
+			return 0, err
+		}
+
+		for _, t := range resp.Topics {
+			if t.Name != topic {
+				log.Debug("unexpected topic information",
+					"expected", topic,
+					"got", t.Name)
+				continue
+			}
+			for _, part := range t.Partitions {
+				if part.ID != partition {
+					log.Debug("unexpected partition information",
+						"topic", t.Name,
+						"expected", partition,
+						"got", part.ID)
+					continue
+				}
+				resErr = part.Err
+
+				switch part.Err {
+				case proto.ErrLeaderNotAvailable, proto.ErrNotLeaderForPartition, proto.ErrBrokerNotAvailable:
+					// Failover happened, so we probably need to talk to a different broker. Let's
+					// kick off a metadata refresh.
+					log.Debug("cannot fetch offset",
```

Comment:
Nit: Would it be helpful to do `log.Debugf("cannot fetch offset for topic %s, partition %d: %v", t.Name, part.ID, part.Err)`?

Also, what's the log level we're using in production?  If it's higher than Debug, would it make sense to raise this to at least Info?